design/clean_artifacts_in_csi_flow.md

```diff
@@ -0,0 +1,487 @@
+		logger.Debugf("VolumeSnapshot or VolumeSnapshotContent are not ready yet.")
+        progress.Description = "VolumeSnapshot or VolumeSnapshotContent creation is still in progress."
+		return progress, nil
+	}
+
+	progress.Description = "VolumeSnapshot and VolumeSnapshotContent are ready"
+	progress.OperationUnits = "Bytes"
+	progress.Completed = true
+	if vsc.Status.RestoreSize != nil {
+		progress.NCompleted = *vsc.Status.RestoreSize
+		progress.NTotal = *vsc.Status.RestoreSize
+	}
+
+	logger.Infof("VolumeSnapshot %s/%s and VolumeSnapshotContent %s are ready.",
+		vs.Namespace, vs.Name, vsc.Name)
+
+	return progress, nil
+}
+```
+
+The VolumeSnapshot and VolumeSnapshotContent are delete in the `restore_finalizer_controller`.
+
+``` go
+func (ctx *finalizerContext) deleteVolumeSnapshotAndVolumeSnapshotContent() (errs results.Result) {
+	for _, operation := range ctx.restoreItemOperationList.items {
+		if operation.Spec.RestoreItemAction == constant.PluginCsiVolumeSnapshotRestoreRIA &&
+			operation.Status.Phase == itemoperation.OperationPhaseCompleted {
+			if operation.Spec.OperationID == "" || !strings.Contains(operation.Spec.OperationID, "/") {
+				ctx.logger.Errorf("invalid OperationID: %s", operation.Spec.OperationID)
+				errs.Add("", errors.Errorf("invalid OperationID: %s", operation.Spec.OperationID))
+				continue
+			}
+
+			operationIDParts := strings.Split(operation.Spec.OperationID, "/")
+
+			vs := new(snapshotv1api.VolumeSnapshot)
+			vsc := new(snapshotv1api.VolumeSnapshotContent)
+
+			if err := ctx.crClient.Get(
+				context.TODO(),
+				client.ObjectKey{Namespace: operationIDParts[0], Name: operationIDParts[1]},
+				vs,
+			); err != nil {
+				ctx.logger.Errorf("Fail to get the VolumeSnapshot %s: %s", operation.Spec.OperationID, err.Error())
+				errs.Add(operationIDParts[0], errors.Errorf("Fail to get the VolumeSnapshot %s: %s", operation.Spec.OperationID, err.Error()))
+				continue
+			}
+
+			if err := ctx.crClient.Delete(context.TODO(), vs); err != nil {
+				ctx.logger.Errorf("Fail to delete VolumeSnapshot %s: %s", operation.Spec.OperationID, err.Error())
+				errs.Add(vs.Namespace, err)
+			}
+
+			if vs.Status != nil && vs.Status.BoundVolumeSnapshotContentName != nil {
+				vsc.Name = *vs.Status.BoundVolumeSnapshotContentName
+			} else {
+				ctx.logger.Errorf("VolumeSnapshotContent %s is not ready.", vsc.Name)
+				errs.Add("", errors.Errorf("VolumeSnapshotContent %s is not ready.", vsc.Name))
+				continue
+			}
+
+			if err := ctx.crClient.Delete(context.TODO(), vsc); err != nil {
+				ctx.logger.Errorf("Fail to delete VolumeSnapshotContent %s: %s", vsc.Name, err.Error())
+				errs.Add("", errors.Errorf("Fail to delete the VolumeSnapshotContent %s", err))
+			}
+		}
+	}
+
+	return errs
+}
+```
+
+### Backup Sync
+csi-volumesnapshotclasses.json, csi-volumesnapshotcontents.json, and csi-volumesnapshots.json are CSI-related metadata files in the BSL for each backup.
```

Comment:
please ensure these files are not removed altogether from velero, we rely on these files for record keeping of snapshots taken. 
Just putting this to make sure the design intent is not to remove these files and only called in doc to explain the need for them in delete/ restore flow

Reply:
OK. Deleting the existing metadata files is considered a breaking change.
Even if this design proposes to delete them, we should follow the deprecation policy.

Reply:
@anshulahuja98 
Apart from whether the deprecation policy is relevant or not, it's OK to not delete the metadata files in v1.16.
Because we also have the Backup's VolumeInfo metadata, I think the downstream consumer, such as Azure, can use that to replace the CSI VSC and VS metadata files.
https://github.com/vmware-tanzu/velero/blob/main/design/Implemented/pv_backup_info.md

What's your opinion on making a change to adopt the backup VolumeInfo instead in the Azure future releases?

Reply:
that seems reasonable, we'll work on migrating to use VolumeInfo instead of CSI JSONs.
We can target removal of files in 1.17 / 1.17+